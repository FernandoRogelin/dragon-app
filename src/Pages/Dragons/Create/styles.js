import styled from "styled-components/macro";

export const Form = styled.form`
  width: 40%;
  padding: 10px 20px;
  border-radius: 15px;
  box-sizing: border-box;
  background-color: ${({ theme }) => theme.white};

  @media only screen and (max-width: 440px) {
    width: 90%;
  }
`;

export const Content = styled.div`
  display: flex;
  margin-bottom: 20px;
  flex-direction: column;

  > input {
    :not(:last-child) {
      margin-bottom: 20px;
    }
  }
`;
