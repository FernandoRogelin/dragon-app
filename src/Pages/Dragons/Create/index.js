import React, { useState } from "react";

import Routes from "../../../Routes";
import { useForm } from "react-hook-form";
import { useHistory } from "react-router-dom";
import { createDragon } from "../../../Services/Dragons";
import { Wrapper, InputForm, Button, Loading } from "../../../Components";

import { Content, Form } from "./styles";

export default function Create() {
  const history = useHistory();
  const [loading, setLoading] = useState(false);

  const { handleSubmit, register, errors, formState } = useForm({
    mode: "onChange",
  });

  async function onSubmit(values) {
    setLoading(true);

    try {
      await createDragon(values);

      history.push(Routes.dragons);
    } catch (error) {
      console.log(error);
    } finally {
      setLoading(false);
    }
  }

  return (
    <Wrapper
      nameButton="Voltar"
      title="Criar novo dragão"
      onClick={() => history.push(Routes.dragons)}
    >
      <Loading isLoading={loading} />
      <Form onSubmit={handleSubmit(onSubmit)} id="formulario">
        <Content>
          <InputForm
            name="name"
            placeholder="Nome"
            error={errors.name}
            testId="input-name"
            ariaLabelledby="formulario nome"
            refInput={register({
              required: "Campo obrigatório.",
            })}
          />
          <InputForm
            name="type"
            placeholder="Tipo"
            error={errors.type}
            testId="input-type"
            ariaLabelledby="formulario tipo"
            refInput={register({
              required: "Campo obrigatório.",
            })}
          />
          <InputForm
            name="createdAt"
            mask="99/99/9999"
            error={errors.createdAt}
            testId="input-createdAt"
            placeholder="Data de criação"
            ariaLabelledby="formulario data"
            refInput={register({
              required: "Campo obrigatório.",
            })}
          />
        </Content>
        <Button
          type="submit"
          aria-label="logar"
          data-testid="button-create"
          disabled={!formState.isValid || formState.isSubmitting}
        >
          Adicionar
        </Button>
      </Form>
    </Wrapper>
  );
}
