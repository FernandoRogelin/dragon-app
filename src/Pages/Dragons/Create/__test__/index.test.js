import React from "react";
import { render, fireEvent, act, wait } from "@testing-library/react";

import Create from "..";

const promise = Promise.resolve();

jest.mock("react-router-dom", () => ({
  useHistory: () => ({
    push: jest.fn(),
  }),
}));

describe("Create dragon component", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test("Snapshot test", async () => {
    const { container } = render(<Create />);

    expect(container.firstChild).toMatchSnapshot();

    await act(() => promise);
  });

  test("Create Dragon", async () => {
    const { getByTestId } = render(<Create />);

    const inputName = getByTestId("input-name");
    const inputType = getByTestId("input-type");
    const inputDate = getByTestId("input-createdAt");

    fireEvent.change(inputName, { target: { value: "Drakar" } });
    fireEvent.change(inputType, { target: { value: "Orange" } });
    fireEvent.change(inputDate, { target: { value: "20072020" } });

    await wait(() => fireEvent.click(getByTestId("button-create")));

    await act(() => promise);
  });
});
