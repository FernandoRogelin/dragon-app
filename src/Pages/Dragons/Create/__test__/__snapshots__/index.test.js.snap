// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create dragon component Snapshot test 1`] = `
<div
  class="styles__WrapperComponent-mms3ln-0 gZmSTy"
>
  <h1
    class="styles__Title-mms3ln-1 wrRRR"
  >
    <button
      class="Button__button-a83c5q-0 kzEeEq"
    >
      Voltar
    </button>
    Criar novo dragão
  </h1>
  <form
    class="styles__Form-adeg7n-0 ffnFVm"
    id="formulario"
  >
    <div
      class="styles__Content-adeg7n-1 jMNlvH"
    >
      <input
        aria-labelledby="formulario nome"
        class="styles__Input-sc-1tlf70t-0 iWnGNM"
        data-testid="input-name"
        name="name"
        placeholder="Nome"
        value=""
      />
      <input
        aria-labelledby="formulario tipo"
        class="styles__Input-sc-1tlf70t-0 iWnGNM"
        data-testid="input-type"
        name="type"
        placeholder="Tipo"
        value=""
      />
      <input
        aria-labelledby="formulario data"
        class="styles__Input-sc-1tlf70t-0 iWnGNM"
        data-testid="input-createdAt"
        name="createdAt"
        placeholder="Data de criação"
        value=""
      />
    </div>
    <button
      aria-label="logar"
      class="Button__button-a83c5q-0 kzEeEq"
      data-testid="button-create"
      type="submit"
    >
      Adicionar
    </button>
  </form>
</div>
`;
